### PR TITLE
feat(node): hide all services behind traits

### PR DESCRIPTION
**Pull request by zvolin**

### CODE REVIEW

File: node/src/p2p.rs [hunk 1]
```diff
@@ -105,65 +146,34 @@ impl P2p {
...
-        self.send_cmd(P2pCmd::NetworkInfo { respond_to: tx })
-            .await?;
-        rx.await
-            .map_err(|_| P2pError::CmdCancelled("NetworkInfo".to_owned()))
-    }
-}
-
-impl Worker {
     async fn run(&mut self) {
         let mut command_stream = self.cmd_rx.clone().into_stream().fuse();
```
Comment (oblique) [inline node/src/p2p.rs:178]: I didn't notice this in the previous PR. I think I prefer `self.cmd_rx.recv_async()`

File: node/src/p2p.rs [hunk 2]
```diff
@@ -66,37 +44,100 @@ pub enum P2pError {
...
+
+#[async_trait]
+pub trait P2pService: Service<Args = P2pArgs, Command = P2pCmd, Error = P2pError> {
+    async fn network_info(&self) -> Result<NetworkInfo> {
+        let (tx, rx) = oneshot::channel();
+
+        self.send_command(P2pCmd::NetworkInfo { respond_to: tx })
+            .await?;
+        rx.await
+            .map_err(|_| P2pError::CmdCancelled("NetworkInfo".to_owned()))
```
Comment (oblique) [inline node/src/p2p.rs:104]: Shouldn't this be `WorkerDied`? I don't see how a command can be cancelled

File: node/src/p2p.rs [hunk 3]
```diff
@@ -66,37 +44,100 @@ pub enum P2pError {
...
+    }
+
+    // async fn subscribe_for_header_sub(&self) -> impl Stream<Item = ExtendedHeader>;
```
Comment (oblique) [inline node/src/p2p.rs:107]: remove this

File: node/src/p2p.rs [hunk 4]
```diff
@@ -66,37 +44,100 @@ pub enum P2pError {
...
+    pub transport: Boxed<(PeerId, StreamMuxerBox)>,
+    pub network_id: String,
+    pub local_keypair: Keypair,
+    pub bootstrap_peers: Vec<Multiaddr>,
+    pub listen_on: Vec<Multiaddr>,
 }
 
 #[derive(Debug)]
-enum P2pCmd {
+pub enum P2pCmd {
```
Comment (oblique) [inline node/src/p2p.rs:61]: I think this shouldn't be public because we will implement wrappers for the commands

File: node/src/syncer.rs [hunk 5]
```diff
@@ -1,16 +1,60 @@
...
+    Service<Args = SyncerArgs<P2pSrv>, Command = SyncerCmd, Error = SyncerError>
+{
+}
+
+#[async_trait]
+impl<P2pSrv: P2pService> SyncerService<P2pSrv> for Syncer<P2pSrv> {}
+
+impl<P2pSrv: P2pService> Syncer<P2pSrv> {
+    pub fn new(store: Arc<RwLock<Store>>, p2p: Arc<P2pSrv>) -> Self {
+        Syncer { store, p2p }
```
Comment (oblique) [inline node/src/syncer.rs:58]: This is not needed because `start` is the constructor

File: node/src/syncer.rs [hunk 6]
```diff
@@ -1,16 +1,60 @@
...
-impl Syncer {
-    pub fn new(store: Arc<RwLock<Store>>) -> Self {
-        Syncer { store }
+pub struct SyncerArgs<P2pSrv: P2pService> {
+    pub p2p: Arc<P2pSrv>,
+    pub store: Arc<RwLock<Store>>,
+}
+
+#[derive(Debug)]
+pub enum SyncerCmd {}
```
Comment (oblique) [inline node/src/syncer.rs:26]: Same as `P2pCmd`, this should be private

File: node/src/p2p.rs (hunk 1 above)
Comment (zvolin) [reply node/src/p2p.rs:178]: I couldn't find any info whether `recv_async` is cancel safe tho

File: node/src/p2p.rs (hunk 4 above)
Comment (zvolin) [reply node/src/p2p.rs:61]: making it private would mean making everything from `Service` trait private too and preventing any external implementations

File: node/src/p2p.rs (hunk 2 above)
Comment (zvolin) [reply node/src/p2p.rs]: removed cmdcancelled entirely

File: node/src/p2p.rs (hunk 4 above)
Comment (oblique) [reply node/src/p2p.rs:58]: That is true, please use `#[doc(hide)]` for now. If we have an issue in the future we may need to wrap it in a struct and have the field as private.

File: node/src/p2p.rs (hunk 1 above)
Comment (oblique) [reply node/src/p2p.rs:172]: It is cancel safe but not documented, we can keep the fused stream for now.

File: node/src/p2p.rs (hunk 4 above)
Comment (zvolin) [reply node/src/p2p.rs:59]: hidden

File: node/src/syncer.rs (hunk 6 above)
Comment (zvolin) [reply node/src/syncer.rs:27]: hidden

**Review by oblique — APPROVED** [review]

**Review by fl0rek — APPROVED** [review]
:guitar:

**Pull request merged**